felix/dataplane/linux/int_dataplane.go

```diff
@@ -1871,6 +1780,127 @@ func (d *InternalDataplane) loopUpdatingDataplane() {
 	}
 }
 
+func newRefreshTicker(name string, interval time.Duration) <-chan time.Time {
+	if interval <= 0 {
+		log.Infof("Refresh of %s on timer disabled", name)
+		return nil
+	}
+	log.WithField("interval", interval).Infof("Will refresh %s on timer", name)
+	refreshTicker := jitter.NewTicker(interval, interval/10)
+	return refreshTicker.Channel()
+}
+
+// onDatastoreMessage is called when we get a message from the calculation graph
+// it opportunistically processes a match of messages from its channel.
+func (d *InternalDataplane) onDatastoreMessage(msg interface{}) {
+	d.datastoreBatchSize = 1
+
+	// Process the message we received, then opportunistically process any other
+	// pending messages.  This helps to avoid doing two dataplane updates in quick
+	// succession (and hence increasing latency) if we're _not_ being throttled.
+	d.processMsgFromCalcGraph(msg)
+	drainChan(d.toDataplane, d.processMsgFromCalcGraph)
+
+	summaryBatchSize.Observe(float64(d.datastoreBatchSize))
+}
+
+func (d *InternalDataplane) processMsgFromCalcGraph(msg interface{}) {
+	log.WithField("msg", proto.MsgStringer{Msg: msg}).Infof("Received %T update from calculation graph", msg)
+	d.datastoreBatchSize++
+	d.dataplaneNeedsSync = true
+	d.recordMsgStat(msg)
+	for _, mgr := range d.allManagers {
+		mgr.OnUpdate(msg)
+	}
+	switch msg.(type) {
+	case *proto.InSync:
+		log.WithField("timeSinceStart", time.Since(processStartTime)).Info(
+			"Datastore in sync, flushing the dataplane for the first time...")
+		d.datastoreInSync = true
+	}
+}
+
+// onIfaceMonitorMessage is called when we get a message from the interface monitor
+// it opportunistically processes a match of messages from its channel.
+func (d *InternalDataplane) onIfaceMonitorMessage(ifaceUpdate any) {
+	// Separate stats for historic reasons: there use to be two channels.
+	d.linkUpdateBatchSize = 0
+	d.addrsUpdateBatchSize = 0
+
+	// As for datastore messages, the interface monitor can send many messages in one go, so we
+	// opportunistically process a batch even if we're not being throttled.
+	d.processIfaceUpdate(ifaceUpdate)
+	drainChan(d.ifaceUpdates, d.processIfaceUpdate)
+
+	d.dataplaneNeedsSync = true
+	if d.linkUpdateBatchSize > 0 {
+		summaryIfaceBatchSize.Observe(float64(d.linkUpdateBatchSize))
+	}
+	if d.addrsUpdateBatchSize > 0 {
+		summaryAddrBatchSize.Observe(float64(d.addrsUpdateBatchSize))
+	}
+}
+
+func (d *InternalDataplane) processIfaceUpdate(ifaceUpdate any) {
+	switch ifaceUpdateMsg := ifaceUpdate.(type) {
+	case *ifaceStateUpdate:
+		d.processIfaceStateUpdate(ifaceUpdateMsg)
+	case *ifaceAddrsUpdate:
+		d.processIfaceAddrsUpdate(ifaceUpdateMsg)
+	case *ifaceInSync:
+		d.processIfaceInSync()
+	}
+}
+
+func (d *InternalDataplane) processIfaceInSync() {
+	if d.ifaceMonitorInSync {
+		return
+	}
+	log.Info("Interface monitor now in sync.")
+	d.ifaceMonitorInSync = true
```

Comment:
does this ever become false? Seems like the iface monitor may set it many times to true

Reply:
No, it doesn't; left over from previous iteration.

Reply:
No, it's a one-way flag; just to let the main loop know that we're in sync at one point.